ts/services/ElementService.ts

```diff
@@ -0,0 +1,185 @@
+import * as Bluebird from 'bluebird';
+import * as fs from 'fs';
+import * as request from 'request-promise';
+
+const MAINT_TAG_NAME = 'n.state.maintenance';
+const MAINT_END_TAG_NAME = 'n.state.maintenance_end';
+const MS_IN_HOUR = 3600000;
+
+class ElementService {
+
+  public async lsMaintenanceMode(config, logger): Promise<void> {
+    logger.info('\nListing elements in maintenance mode');
+    try {
+      const response = await request({
+        auth: {
+          pass: config.password,
+          user: config.username
+        },
+        body: {
+          elementTags: {
+            and: true,
+            items: [{
+              contains: true,
+              key: MAINT_TAG_NAME,
+              literal: true,
+              value: 'true'
+            }]
+          },
+          page: 0,
+          pageSize: 50,
+          sort: {
+            field: 'name',
+            missing: '_last',
+            order: 'asc'
+          },
+          sourceFilter: {
+            excludes: ['metrics']
+          }
+        },
+        json: true,
+        method: 'POST',
+        uri: `${config.endpoint}/elements/elasticsearch/elementQuery`
+      });
+      if (config.format === 'text') {
+        logger.info(response.page.content.map((el) => {
+          const endEpoch = el.netuitiveTags[MAINT_END_TAG_NAME];
+          if (typeof endEpoch === 'undefined' || endEpoch < 0) {
+            return el.name + ' id(' + el.id + ')' + ' ending(never)';
+          } else {
+            const endDate = new Date( parseInt(endEpoch, 10) );
+            return `${el.name} id(${el.id}) ending(${endDate.toUTCString()})`;
+          }
+        }));
+      }
+      if (config.format === 'json') {
+        logger.info(JSON.stringify(response, null, 2));
+      }
+    } catch (e) {
+      logger.error('There was an error listing the elements: ' + e);
+    }
+  }
+
+  public async startMaintenanceMode(id, config, logger): Promise<void> {
```

Comment:
This logic seems odd to have in the CLI. Getting tags, updating/creating them, and using two conventional tags feels overly complicated, not to mention it will be difficult to document.

I thought there was going to be a new API endpoint which did all of this behind the scenes. Otherwise all this logic will be duplicated within the UI.

Reply:
We discussed this... we are going with the current impl